falcon/api.py

```diff
@@ -544,13 +545,13 @@ def _call_req_mw(self, stack, req, resp):
             # Put executed component on the stack
             stack.append(component)  # keep track from outside
 
-    def _call_rsrc_mw(self, stack, req, resp, resource):
+    def _call_rsrc_mw(self, stack, req, resp, resource, params):
         """Run process_resource middleware methods."""
 
         for component in self._middleware:
             _, process_resource, _ = component
             if process_resource is not None:
-                process_resource(req, resp, resource)
+                process_resource(req, resp, resource, params)
```

Comment:
One way to mitigate the performance issue is to do the work up front in `prepare_middleware` to detect the number of args, and put a shim in place as needed for older middleware components. E.g. (pseudocode):

```
if is old-style method:
    @functools.wraps(orig_process_resource)
    def func(req, resp, resource, params)):
         orig_process_resource(req, resp, resource)

    process_resource = func
else:
    process_resource = orig_process_resource
```

Then the overhead is reduced to an extra function call. Eventually we can remove the shim as a breaking change once everyone has had a chance to migrate (e.g., 2.0).

What do you think?